security/advancedtls/advancedtls_integration_test.go

```diff
@@ -408,3 +411,272 @@ func (s) TestEnd2End(t *testing.T) {
+			if err != nil {
+				t.Fatalf("copyFileContents(%s, %s) failed: %v", testdata.Path("server_trust_cert_1.pem"), serverTrustTmp.Name(), err)
+			}
+			// Create PEMFileProvider(s) watching the content changes of temporary
+			// files.
+			clientIdentityProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				CertFile:         clientCertTmp.Name(),
+				KeyFile:          clientKeyTmp.Name(),
+				IdentityInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer clientIdentityProvider.Close()
+			clientRootProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				TrustFile:    clientTrustTmp.Name(),
+				RootInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer clientRootProvider.Close()
+			serverIdentityProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				CertFile:         serverCertTmp.Name(),
+				KeyFile:          serverKeyTmp.Name(),
+				IdentityInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer serverIdentityProvider.Close()
+			serverRootProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				TrustFile:    serverTrustTmp.Name(),
+				RootInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer serverRootProvider.Close()
+			// Start a server and create a client using advancedtls API with Provider.
+			serverOptions := &ServerOptions{
+				IdentityOptions: IdentityCertificateOptions{
+					IdentityProvider: serverIdentityProvider,
+				},
+				RootOptions: RootCertificateOptions{
+					RootProvider: serverRootProvider,
+				},
+				RequireClientCert: true,
+				VerifyPeer: func(params *VerificationFuncParams) (*VerificationResults, error) {
+					return &VerificationResults{}, nil
+				},
+				VType: CertVerification,
+			}
+			serverTLSCreds, err := NewServerCreds(serverOptions)
+			if err != nil {
+				t.Fatalf("failed to create server creds: %v", err)
+			}
+			s := grpc.NewServer(grpc.Creds(serverTLSCreds))
+			defer s.Stop()
+			lis, err := net.Listen("tcp", port)
+			if err != nil {
+				t.Fatalf("failed to listen: %v", err)
+			}
+			defer lis.Close()
+			pb.RegisterGreeterServer(s, greeterServer{})
+			go s.Serve(lis)
+			clientOptions := &ClientOptions{
+				IdentityOptions: IdentityCertificateOptions{
+					IdentityProvider: clientIdentityProvider,
+				},
+				VerifyPeer: func(params *VerificationFuncParams) (*VerificationResults, error) {
+					return &VerificationResults{}, nil
+				},
+				RootOptions: RootCertificateOptions{
+					RootProvider: clientRootProvider,
+				},
+				VType: CertVerification,
+			}
+			clientTLSCreds, err := NewClientCreds(clientOptions)
+			if err != nil {
+				t.Fatalf("clientTLSCreds failed to create, error: %v", err)
+			}
```

Comment:
Upto this point seems to be setup. So, please make a helper function which is clearly documented as to what it does and just invoke that from here so that people reading the test can get to the meat of it easily.

Reply:
Sure, I added several helper functions. Hope it will make the code cleaner.